qp_shotgun/kneaddata/kneaddata.py

```diff
@@ -0,0 +1,272 @@
+                                 '%s' % f_fn)
+
+        # make sure that we got one matching run prefix:
+        if run_prefix is None:
+            raise ValueError('No run prefix matching this fwd read: %s'
+                             % f_fn)
+
+        if run_prefix in used_prefixes:
+            raise ValueError('This run prefix matches multiple fwd reads: '
+                             '%s' % run_prefix)
+
+        # if we have reverse reads, make sure the matching pair also
+        # matches the run prefix:
+        if (reverse_seqs and not
+                basename(reverse_seqs[i]).startswith(run_prefix)):
+            raise ValueError('Reverse read does not match this run prefix. '
+                             'Run prefix: %s\nForward read: %s\n'
+                             'Reverse read: %s\n' %
+                             (run_prefix, f_fn, basename(reverse_seqs[i])))
+
+        used_prefixes.append(run_prefix)
+        # create the tuple for this read set
+        if reverse_seqs:
+            samples.append((run_prefix, sn_by_rp[run_prefix], f_fp,
+                            reverse_seqs[i]))
+        else:
+            samples.append((run_prefix, sn_by_rp[run_prefix], f_fp, None))
+
+    return(samples)
+
+
+def format_kneaddata_params(parameters):
+
+    params = []
+
+    for param in sorted(parameters):
+        value = parameters[param]
+
+        if str(value) == 'True':
+            params.append('--%s' % param)
+            continue
+        elif str(value) == 'False':
+            continue
+        elif value:
+            params.append('--%s %s' % (param, value))
+            continue
+
+    param_string = ' '.join(params)
+
+    return(param_string)
+
+
+def generate_kneaddata_commands(forward_seqs, reverse_seqs, map_file,
+                                out_dir, parameters):
+    """Generates the KneadData commands
+
+    Parameters
+    ----------
+    forward_seqs : list of str
+        The list of forward seqs filepaths
+    reverse_seqs : list of str
+        The list of reverse seqs filepaths
+    map_file : str
+        The path to the mapping file
+    out_dir : str
+        The job output directory
+    parameters : dict
+        The command's parameters, keyed by parameter name
+
+    Returns
+    -------
+    list of str
+        The KneadData commands
+    list of str
+        The run prefixes used
+
+
+    Raises
+    ------
+    ValueError
+        If the rev is not an empty list and the same length than fwd seqs
+        The prefixes of the run_prefix don't match the file names
+
+    Notes
+    -----
+    """
+    # making sure the forward and reverse reads are in the same order
+
+    # we match filenames, samples, and run prefixes
+    samples = make_read_pairs_per_sample(forward_seqs, reverse_seqs, map_file)
+
+    cmds = []
+
+    param_string = format_kneaddata_params(parameters)
+    prefixes = []
+    for run_prefix, sample, f_fp, r_fp in samples:
```

Comment:
Yes, and it's a big pain in the ass. This frequently happens with shallow shotgun runs like we have been getting a lot of recently, where the quality is not so good and you end up with the reverse read being trimmed to extinction. 

I agree that this kind of thing can set us up for problems and we should keep aware of it! Perhaps for early QC tools like kneaddata we can enforce proper paired reads (assuming no trimming so everything should be consistent) while downstream tools can be more flexible?